temporal-sdk/src/main/java/io/temporal/internal/worker/LocalActivityWorker.java

```diff
@@ -256,51 +263,70 @@ public boolean dispatch(
     private boolean submitANewExecution(
         @Nonnull LocalActivityExecutionContext executionContext,
         @Nonnull PollActivityTaskQueueResponse.Builder activityTask,
-        @Nullable Deadline acceptanceDeadline) {
+        @Nullable Deadline heartbeatDeadline) {
+      long acceptanceTimeoutMs = 0;
+      boolean timeoutIsScheduleToStart = false;
+      if (heartbeatDeadline != null) {
+        acceptanceTimeoutMs = heartbeatDeadline.timeRemaining(TimeUnit.MILLISECONDS);
+      }
+      Duration scheduleToStartTimeout = executionContext.getScheduleToStartTimeout();
+      if (scheduleToStartTimeout != null) {
+        long scheduleToStartTimeoutMs = scheduleToStartTimeout.toMillis();
+        if (scheduleToStartTimeoutMs > 0 && scheduleToStartTimeoutMs < acceptanceTimeoutMs) {
+          acceptanceTimeoutMs = scheduleToStartTimeoutMs;
+          timeoutIsScheduleToStart = true;
+        }
+      }
       try {
-        boolean accepted;
-        if (acceptanceDeadline == null) {
-          newExecutionsBackpressureSemaphore.acquire();
-          accepted = true;
+        SlotPermit permit = null;
+        SlotReservationData reservationCtx =
+            new SlotReservationData(taskQueue, options.getIdentity(), options.getBuildId());
+        if (acceptanceTimeoutMs <= 0) {
+          permit = slotSupplier.reserveSlot(reservationCtx);
         } else {
-          long acceptanceTimeoutMs = acceptanceDeadline.timeRemaining(TimeUnit.MILLISECONDS);
-          if (acceptanceTimeoutMs > 0) {
-            accepted =
-                newExecutionsBackpressureSemaphore.tryAcquire(
-                    acceptanceTimeoutMs, TimeUnit.MILLISECONDS);
-          } else {
-            accepted = newExecutionsBackpressureSemaphore.tryAcquire();
-          }
-          if (!accepted) {
-            log.warn(
-                "LocalActivity queue is full and submitting timed out for activity {} with acceptanceTimeoutMs: {}",
-                activityTask.getActivityId(),
-                acceptanceTimeoutMs);
+          try {
+            TimeLimiter timeLimiter = SimpleTimeLimiter.create(timeoutThreadPool);
+            permit =
+                timeLimiter.callWithTimeout(
```

Comment:
This is the actual fix. The source of the bug was not assigning the permit when reservation was called behind this `callWithTimeout` function.